### ISSUE TITLE
Reimplementing the wheel?

### ISSUE THREAD

**Issue by qpippop** [ivanpaulovich/FluentMediator#47]
I was digging through the source code and noticed that IPipeline is very similar to TPL Dataflow block. The IPipeline has almost the same (although rather limited) functionality.

```c#
IPipelineBuilder<TRequest>.Call<THandler> // new ActionBlock<TRequest>(handlerAction)
IPipelineBuilder<TRequest>.Return<TResult, THandler> // new TransformBlock<TRequest, TResult>(handlerFunc)
IPipeline.Publish(GetService getService, object request) // IPropagatorBlock<object, object>.Post(request)
IPipeline.Send<TResult>(GetService getService, object request) // IPropagatorBlock<object, object>.Post(request)
                                                               // (TResult)IPropagatorBlock<object, object>.Receive();
```
At first glance, there is a small problem here: you need to find a way to pass THandler into dataflow block processing method (and I'm not sure at what level of abstraction this should happen).
If this can be bypassed, the pipes construction will look like this
```c#
private readonly IPropagatorBlock<TRequest, TRequest> _propagatorBlock;
public PipelineBuilder(string? name)
{
    _propagatorBlock = new BufferBlock<TRequest>();
    _name = name;
}
// ...
// do something with THandler
public IPipelineBuilder<TRequest> Call<THandler>(Action<TRequest> action) // signature will change probably because of remark above
{
    // we can pass ExecutionDataflowBlockOptions to specify bounded capacity, max degree of parallelism etc.
    var nextBlock = new ActionBlock<TRequest>(action);
    _propagatorBlock.LinkTo(nextBlock /*, predicate for TRequest */);

    return this;
}

// do something with THandler
public IPipeline Return<TResult, THandler>(Func<TRequest, TResult> func)
{
    // we can pass ExecutionDataflowBlockOptions to specify bounded capacity, max degree of parallelism etc.
    var nextBlock = new TransformBlock<TRequest, TResult>(func);
    _propagatorBlock.LinkTo(nextBlock);

    return Build();
}
```
The main point here - TPL Dataflow is the mature lib which can be used as pipe backend.

**Comment by ivanpaulovich** [comment]
Hey @qpippop,

I did not know [TPL](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/dataflow-task-parallel-library) had this similar functionality.

It is very interesting how you made the comparison! I just learned something new 🍺 

As you said it the TPL could become the pipe backend. At this moment I need to learn about it, maybe I am gonna use the next holidays.

**Comment by ivanpaulovich** [comment]
Hey @qpippop,

I spent some time with TPL to see if I could start using it on FluentMediator.

I found a design issue that I couldn't solve. 

The `ActionBlock<TRequest>` accepts only one generic type argument which is not enough for our needs. I wish ActionBlock had a signature  `ActionBlock<THandler, TRequest>`  so I could design the pipeline like:

```cs
m.On<PingRequest>().Pipeline()
    .Call<IPingHandler>((handler, req) => handler.MyCustomFooMethod(req))
    .Call<IPingHandler>((handler, req) => handler.MyCustomBarMethod(req));
```

Anyway, it was an interesting input you brought to us.
